tensorflow/compiler/mlir/tensorflow/transforms/convert_tf_control_flow_to_scf.cc

```diff
@@ -98,11 +103,86 @@ class ConvertIfRegionOp : public OpRewritePattern<IfRegionOp> {
   }
 };
 
+/// Convert the `tf.WhileRegion` op to the `scf.while` op iff the condition of
+/// the `tf.WhileRegion` op is a 0-D tensor of 1-bit signless integers.
+class ConvertWhileRegionOp : public OpRewritePattern<WhileRegionOp> {
+ public:
+  using OpRewritePattern<WhileRegionOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(WhileRegionOp op,
+                                PatternRewriter& rewriter) const override {
+    // Creates the `before` or `after` region of the `scf.while` op. Note that
+    // `tf_cond_or_body_region` is the `cond` or `body` region of the
+    // `tf.WhileRegion` op. `scf_before_or_after_region` is the `before` or
+    // `after` region of the new `scf.while` op.
+    auto createScfCondOrBody = [](Region& tf_cond_or_body_region,
+                                  Region& scf_before_or_after_region,
+                                  PatternRewriter& rewriter) {
+      Block* first_block_of_scf_before_or_after_region =
+          cloneRegionAndGetFirstBlock(tf_cond_or_body_region,
+                                      scf_before_or_after_region, rewriter);
+
+      Operation* cond_or_body_terminator =
+          first_block_of_scf_before_or_after_region->getTerminator();
+      rewriter.setInsertionPoint(cond_or_body_terminator);
+      return cond_or_body_terminator;
+    };
+
+    Region& tf_cond_region = op.cond();
+    Region& tf_body_region = op.body();
+
+    auto tf_condition_ranked_type = tf_cond_region.front()
+                                        .getTerminator()
+                                        ->getOperand(0)
+                                        .getType()
+                                        .dyn_cast<RankedTensorType>();
+    // If the condition is unranked, do nothing.
+    if (!tf_condition_ranked_type) return failure();
+    // If the condition is not 0-D, do nothing.
+    if (tf_condition_ranked_type.getRank() != 0) return failure();
+    // If the type of values in the condition tensor is not `1-bit signless
+    // integer`, do nothing.
+    if (!getElementTypeOrSelf(tf_condition_ranked_type).isInteger(1))
+      return failure();
```

Comment:
It isn't clear to me that these conditions are catching legal situation? Are you able to exercise these with a test?

I see in the verifier:
```
  auto cond_type =
      cond_yield->getOperand(0).getType().dyn_cast<RankedTensorType>();
  if (!cond_type || !cond_type.getShape().equals({}) ||
      !cond_type.getElementType().isInteger(/*width=*/1))
    return op.emitOpError()
           << "condition should have a single tensor<i1> result";
```

Reply:
Yes, this wasn't required. Removed it.